openmls/src/binary_tree/array_representation/diff.rs

```diff
@@ -0,0 +1,602 @@
+//! This module provides the `AbDiff` and `StagedAbDiff` structs that allow
+//! performing changes to an `ABinaryTree` instance without immediately applying
+//! the them. Instead, the changes can be applied to the diff and the results
+//! examined before merging the given diff back into the tree (or not).
+//!
+//! ### Don't Panic!
+//!
+//! Functions in this module should never panic. However, if there is a bug in
+//! the implementation, a function will return an unrecoverable `LibraryError`.
+//! This means that some functions that are not expected to fail and throw an
+//! error, will still return a `Result` since they may throw a `LibraryError`.
+
+use std::collections::BTreeMap;
+
+use std::fmt::Debug;
+
+use crate::binary_tree::{array_representation::treemath::sibling, LeafIndex, TreeSize};
+
+use super::{
+    tree::{to_node_index, ABinaryTree, ABinaryTreeError, NodeIndex},
+    treemath::{direct_path, left, lowest_common_ancestor, right, root, TreeMathError},
+};
+
+/// The `StagedAbDiff` can be created from an `AbDiff` instance. It's sole
+/// purpose is to be subsequently merged into an existing `ABinaryTree`
+/// instance. The difference between `StagedAbDiff` and an `AbDiff` is that a
+/// `StagedAbDiff` is immutable and does not contain a reference to the original
+/// tree. Since it only contains the actual diff without reference to the
+/// original content, it can't provide the same information as the `AbDiff` it
+/// was created from. However, the lack of the internal reference means that its
+/// lifetime is not tied to that of the original tree.
+#[derive(Debug)]
+pub(crate) struct StagedAbDiff<T: Clone + Debug> {
+    diff: BTreeMap<NodeIndex, T>,
+    size: TreeSize,
+}
+
+impl<'a, T: Clone + Debug> From<AbDiff<'a, T>> for StagedAbDiff<T> {
+    fn from(diff: AbDiff<'a, T>) -> Self {
+        StagedAbDiff {
+            diff: diff.diff,
+            size: diff.size,
+        }
+    }
+}
+
+impl<T: Clone + Debug> StagedAbDiff<T> {
+    pub(super) fn diff(self) -> BTreeMap<NodeIndex, T> {
+        self.diff
+    }
+
+    pub(super) fn size(&self) -> TreeSize {
+        self.size
+    }
+}
+
+/// A `NodeReference` represents the position of a node in an `AbDiff`. It can
+/// be used to access the node at that position or to navigate to other,
+/// neighbouring nodes via the `sibling`, `left_child` and `right_child`
+/// functions of the `AbDiff`.
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct NodeReference {
+    node_index: NodeIndex,
+}
+
+impl NodeReference {
+    /// Creates a new `NodeReference` to a node at the given index. Returns an
```

Comment:
Because we've been talking about comment guidelines.
The "Return" part of a comment should always be in a new line (probably separated by a new line) to make it easier to find.